storage: TestNodeLivenessStatusMap failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/7ce0cba090528e7c50bd8fe708d536362f5d766a

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestNodeLivenessStatusMap PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=977385&tab=buildLog

```
I181020 05:36:37.766818 106689 storage/node_liveness_test.go:833  [in test] shutting down node 2
I181020 05:36:37.766847 106689 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181020 05:36:37.767103 107665 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: initial connection heartbeat failed: rpc error: code = Canceled desc = context canceled
I181020 05:36:37.767356 106689 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
I181020 05:36:37.768339 106689 storage/node_liveness_test.go:835  [in test] done shutting down node 2
I181020 05:36:37.768363 106689 storage/node_liveness_test.go:838  [in test] decommissioning node 3
I181020 05:36:37.772497 107922 server/server_update.go:64  [n4] failed attempt to upgrade cluster version, error: node 4 not running (UNKNOWN), cannot determine version
I181020 05:36:37.772824 106689 sql/event_log.go:126  [in test] Event: "node_decommissioned", target: 3, info: {}
I181020 05:36:37.772858 106689 storage/node_liveness_test.go:842  [in test] done decommissioning node 3
I181020 05:36:37.772899 106689 storage/node_liveness_test.go:845  [in test] decommissioning and shutting down node 4
I181020 05:36:37.773066 107805 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] kv.TxnCoordSender: heartbeat loop
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
I181020 05:36:37.773087 107804 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181020 05:36:37.773260 107802 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181020 05:36:37.773390 107805 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
I181020 05:36:37.773430 107805 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] closedts-rangefeed-subscriber
I181020 05:36:37.781078 107805 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
I181020 05:36:37.782268 107804 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I181020 05:36:37.785462 107802 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
W181020 05:36:37.785680 107924 internal/client/txn.go:532  [n4] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
W181020 05:36:37.785706 107924 server/node.go:942  [n4] node=4: unable to log node_join event: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
W181020 05:36:37.785815 107909 internal/client/txn.go:532  [n4,hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
I181020 05:36:37.785845 107909 storage/node_liveness.go:817  [n4,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted])
W181020 05:36:37.786022 107909 internal/client/txn.go:532  [n4,hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W181020 05:36:37.786045 107909 storage/node_liveness.go:494  [n4,hb] failed node liveness heartbeat: node unavailable; try another peer
I181020 05:36:37.794717 107745 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
I181020 05:36:37.794775 107744 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n1: context canceled
W181020 05:36:37.800053 107903 ts/db.go:195  [n4,ts-poll] error writing time series data: node unavailable; try another peer
--- FAIL: TestNodeLivenessStatusMap (0.30s)
	node_liveness_test.go:847: during liveness update 4 -> true: node not in the liveness table

```

--- a/pkg/storage/node_liveness_test.go
+++ b/pkg/storage/node_liveness_test.go
@@ -824,6 +824,7 @@ func TestNodeLivenessStatusMap(t *testing.T) {
 
 	log.Infof(ctx, "waiting for node statuses")
 	tc.WaitForNodeStatuses(t)
+	tc.WaitForNodeLiveness(t)
 	log.Infof(ctx, "waiting done")
 
 	firstServer := tc.Server(0).(*server.TestServer)

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -33,6 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
+	"github.com/cockroachdb/cockroach/pkg/storage/storagepb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -592,6 +593,26 @@ func (tc *TestCluster) WaitForNodeStatuses(t testing.TB) {
 			if len(node.StoreStatuses) == 0 {
 				return fmt.Errorf("missing StoreStatuses in NodeStatus: %+v", node)
 			}
+		}
+		return nil
+	})
+}
+
+// WaitForNodeLiveness waits until a liveness record is persisted for every
+// node in the cluster.
+func (tc *TestCluster) WaitForNodeLiveness(t testing.TB) {
+	testutils.SucceedsSoon(t, func() error {
+		db := tc.Servers[0].DB()
+		for _, s := range tc.Servers {
+			key := keys.NodeLivenessKey(s.NodeID())
+			var liveness storagepb.Liveness
+			if err := db.GetProto(context.Background(), key, &liveness); err != nil {
+				return err
+			}
+			if (liveness == storagepb.Liveness{}) {
+				return fmt.Errorf("no liveness record")
+			}
+			fmt.Printf("n%d: found liveness\n", s.NodeID())
 		}
 		return nil
 	})

